### PR TITLE
Implementation of Flow Arcade Token, scripts and initial tests.

### PR DESCRIPTION
**Pull request by robmyers**
* Code is based on Flow Token. Transaction parameters have been re-ordered in some places to be more regular

* Tests are unstructured and will be improved after initial review.

### CODE REVIEW

File: contracts/FlowArcadeToken.cdc
Comment (10thfloor) [inline contracts/FlowArcadeToken.cdc:33]:
```suggestion
    // The functions in the Vault are governed by the pre and post conditions
```

File: contracts/FlowArcadeToken.cdc
Comment (10thfloor) [inline contracts/FlowArcadeToken.cdc:94]: Should return type be `@FlowArcadeToken.Vault`?

**Comment by turbolent** [comment]
@robmyers 

> Transaction parameters have been re-ordered in some places to be more regular

Can you elaborate a bit, it's hard to see in the PR how it differs.
For consistency reasons I'd suggest keeping them the same across different contracts to avoid developer friction: when preparing a transaction the arguments for it have to be added in the correct order – if there's a difference, this might lead to confusion, bugs, etc.

File: transactions/flowArcadeToken/admin/add_minter.cdc
Comment (turbolent) [inline transactions/flowArcadeToken/admin/add_minter.cdc:19]: Maybe consider using type inference to make the code easier to read:

```suggestion
        // Create a reference to the admin admin resource in storage
       let tokenAdmin = adminAcct.borrow<&FlowArcadeToken.Administrator>(from: /storage/flowArcadeTokenAdmin)
            ?? panic("Could not borrow a reference to the admin resource")

        // Create a new minter and a private link to it
        let minter <- tokenAdmin.createNewMinter()
        adminAcct.save(<- minter, to: /storage/flowArcadeTokenMinter)
        let minterCap = adminAcct.link<&FlowArcadeToken.Minter>(
            /private/flowArcadeTokenMinter,
            target: /storage/flowArcadeTokenMinter
        )!

        // Transfer the capability to the minter account
        minterAcct.save(minterCap, to: /storage/flowArcadeTokenMinter)
```

File: contracts/FlowArcadeToken.cdc
Comment (robmyers) [reply contracts/FlowArcadeToken.cdc:94]: Assuming FlowToken, which returns the same type for the same function, is correct this is the proper type.

File: contracts/FlowArcadeToken.cdc
Comment (robmyers) [reply contracts/FlowArcadeToken.cdc:33]: Fixed. Thank you!

File: transactions/flowArcadeToken/admin/add_minter.cdc
Comment (robmyers) [reply transactions/flowArcadeToken/admin/add_minter.cdc:19]: Yes that is much clearer, thank you! The verbose style was for my own learning.

File: contracts/FlowArcadeToken.cdc
Comment (turbolent) [reply contracts/FlowArcadeToken.cdc:94]: I think @10thfloor point was to change the return type to the concrete type. This was originally not possible in earlier versions of Cadence, but is supported now.

File: contracts/FlowArcadeToken.cdc
Comment (turbolent) [inline contracts/FlowArcadeToken.cdc:28]: Why is this commented out?

File: transactions/flowArcadeToken/setup_account.cdc
Comment (turbolent) [inline transactions/flowArcadeToken/setup_account.cdc:30]: Consider using `check` instead of `borrow`, and maybe consider returning early to avoid deep nesting:
```suggestion
        if signer.check<&FlowArcadeToken.Vault>(from: /storage/flowArcadeTokenVault) {
            return
        }
        
        // Create a new FlowArcadeToken Vault and put it in storage
        signer.save(<-FlowArcadeToken.createEmptyVault(), to: /storage/flowArcadeTokenVault)

        // Create a public capability to the Vault that only exposes
        // the deposit function through the Receiver interface
        signer.link<&FlowArcadeToken.Vault{FungibleToken.Receiver}>(
            /public/flowArcadeTokenReceiver,
            target: /storage/flowArcadeTokenVault
        )

        // Create a public capability to the Vault that only exposes
        // the balance field through the Balance interface
        signer.link<&FlowArcadeToken.Vault{FungibleToken.Balance}>(
            /public/flowArcadeTokenBalance,
            target: /storage/flowArcadeTokenVault
        )
    }
}
```

File: transactions/flowArcadeToken/mint_tokens.cdc
Comment (turbolent) [inline transactions/flowArcadeToken/mint_tokens.cdc:16]: Minor: maybe consider indenting this to make it easier to read:

```suggestion
        self.tokenMinter = minterAccount
            .borrow<&FlowArcadeToken.Minter>(from: /storage/flowArcadeTokenMinter) 
            ?? panic("No minter available")

        self.tokenReceiver = getAccount(recipientAddress)
            .getCapability(/public/flowArcadeTokenReceiver)!
            .borrow<&{FungibleToken.Receiver}>()
            ?? panic("Unable to borrow receiver reference")
```

**Comment by robmyers** [comment]
> @robmyers
> 
> > Transaction parameters have been re-ordered in some places to be more regular
> 
> Can you elaborate a bit, it's hard to see in the PR how it differs.
> For consistency reasons I'd suggest keeping them the same across different contracts to avoid developer friction: when preparing a transaction the arguments for it have to be added in the correct order – if there's a difference, this might lead to confusion, bugs, etc.

The parameters in transactions/flowToken/transfer_tokens.cdc are reversed relative to transactions/flowToken/mint_tokens.cdc . My own preference for parameters is "objects first then simple types, large to small, subject to others" and mint_tokens.cdc's `recipientAddress: Address, amount: UFix64` fits that better so I modified  transactions/flowArcadeToken/transfer_tokens.cdc to fit that instead.

I've reverted it so that the transactions have the same interface for each token.

Thank you!

File: contracts/FlowArcadeToken.cdc
Comment (robmyers) [reply contracts/FlowArcadeToken.cdc:28]: We do not currently support burning.

**Comment by turbolent** [comment]
@robmyers Ah, thanks for explaining! Yeah, ideally it would be consistent even in the FT code itself. Maybe we can still fix it there @joshuahannan?

File: contracts/FlowArcadeToken.cdc
Comment (turbolent) [reply contracts/FlowArcadeToken.cdc:28]: Then I'd recommend removing it

File: contracts/FlowArcadeToken.cdc
Comment (robmyers) [reply contracts/FlowArcadeToken.cdc:94]: This would result in different behaviour from FlowToken. I'm happier returning the concrete type if it is OK to do so. :-)

File: contracts/FlowArcadeToken.cdc
Comment (robmyers) [reply contracts/FlowArcadeToken.cdc:28]: Done! Thank you.

File: contracts/FlowArcadeToken.cdc
Comment (robmyers) [reply contracts/FlowArcadeToken.cdc:94]: I've changed it to use the concrete type. Thank you!

File: lib/go/templates/flow_arcade_token_templates.go
Comment (joshuahannan) [inline lib/go/templates/flow_arcade_token_templates.go:24]:
```suggestion
func replaceFATAddress(code, fatAddr string) string {
```

we already have the `replaceAddresses` function that replaces fungible token and flow token addresses, so we should probably rename this and make it only replace the FAT address

Comment (joshuahannan) [reply lib/go/templates/flow_arcade_token_templates.go:24]: And change all the templates in the file to call both replace functions. Do you think that sounds ok?

File: transactions/flowArcadeToken/admin/add_minter.cdc
Comment (joshuahannan) [inline transactions/flowArcadeToken/admin/add_minter.cdc:6]:
```suggestion
        // Create a reference to the admin resource in storage
```

File: lib/go/test/flow_arcade_token_test.go
Comment (joshuahannan) [inline lib/go/test/flow_arcade_token_test.go:382]:
```suggestion
	t.Run("Account with insufficient tokens should not be able to transfer a larger amount", func(t *testing.T) {
```

**Review by joshuahannan — APPROVED** [review]
This looks really good! I just had a few comments.

I like what you're doing with making the tests more modular. We definitely should have a conversation soon about how we can standardize that.

File: lib/go/templates/flow_arcade_token_templates.go
Comment (robmyers) [reply lib/go/templates/flow_arcade_token_templates.go:24]: Yes that makes sense. I'll make that change

File: transactions/flowArcadeToken/admin/add_minter.cdc
Comment (robmyers) [reply transactions/flowArcadeToken/admin/add_minter.cdc:6]: Thank you for catching this!

File: contracts/FlowArcadeToken.cdc
Comment (robmyers) [reply contracts/FlowArcadeToken.cdc:5]: Thank you for catching this!

**Comment by robmyers** [comment]
I recommend rebuilding the contract and transaction resources after merging this.

**Comment by robmyers** [comment]
DO NOT MERGE.

There is a critical issue with the minter and the tests for it. Fixing now...

**Comment by joshuahannan** [comment]
What was the issue @robmyers ?

**Comment by robmyers** [comment]
> What was the issue @robmyers ?

@joshuahannan The FAT implementation of minting, in which a Capability for minting is deposited into the minter account's storage, is naive and attempts to use a capability from one account directly in another from /storage/. This always returns null for the capability.

This was concealed because both the resource and the capability have the same path in their respective accounts' storage and in the tests the minter was the admin due to the wrong account details being used to mint tokens. So it could borrow the resource (rather than the capability) and mint.

Using the minter's account details to mint with reveals the problem. This is very basic Cadence but I made this error at precisely the wrong point in my personal learning curve.

I think I need to wrap the minter capability in another resource and deposit *that* in the minter's storage. This wrapping of a capability in a resource given to another account will be like TokenForwarding.cdc in the FT.

The reason for the gymnastics around minting is that the capacity of a FAT minter is unlimited, so I feel they must be unilaterally revokable by us.

**Comment by robmyers** [comment]
> DO NOT MERGE.
> 
> There is a critical issue with the minter and the tests for it. Fixing now...

Fixed. The Minter capability now works properly, with just one extra layer of indirection as per the programming axiom.

**Pull request closed without merging**
